### ISSUE TITLE
Read output timeouts can trigger early when no output has been written to stream

### ISSUE THREAD

**Issue by yuki-nishida-exa** [ParallelSSH/parallel-ssh#180]
Hello, I have trouble with timeout for `run_command` as it kicks in too early when stderr gets output earlier than stdout.

__Steps to reproduce:__
1. Example code that produces error.
I have Python script like below:
```
client = ParallelSSHClient([<host where script runs>], pool_size=100, timeout=30)
output = client.run_command('/tmp/test_script.sh', use_pty=False, timeout=10, sudo=False)

for host, host_out in output.items():
        for stream in ['stdout', 'stderr']:
            print('Read {}'.format(stream))
            try:
                for line in getattr(host_out, stream):
                    print(line)
            except Timeout:
                print('{} timed out'.format(stream))

client.join(output)
```
`/tmp/test_script.sh` is below:
```
#!/bin/bash

sleep 1
echo "start" >&2   # Redirection affected the behavior 
for i in 1 5 10 15 20 ; do
  sleep $i
  echo $i
done
```

2. Stack trace or error messages.
```
Traceback (most recent call last):
...
  File <my python code>,
    for line in getattr(result, stream):
  File "/opt/exabeam_installer/.env/lib/python2.7/site-packages/pssh/clients/native/single.py", line 401, in read_output_buffer
    for line in output_buffer:
  File "pssh/native/_ssh2.pyx", line 47, in _read_output
pssh.exceptions.Timeout
```

__Expected behaviour:__
stdout should get a few lines before hitting 10 secs timeout as below, regardless of redirecting first output to stderr in `echo "start" >&2`.
```
Read stdout
1
5
10
stdout timed out
Read stderr
start
stderr timed out
```

__Actual behaviour:__
Reading stdout timed out without getting anything as below. 
```
Read stdout
stdout timed out
Read stderr
start
stderr timed out
```

As I experimented without redirection in echoing "start", stdout was retrieved with the lines (1, 5, and 10) as the expected behavior.

__Additional info:__
* Host OS: CentOS7
* Python version: 2.7.5 (one bundled in CentOS 7)
* parallel-ssh version: 1.9.1
* ssh2-python version: 0.18.0.post1
* Note: I originally commented in #109, as I had issue with running a hadoop command and the phenomena looked similar. But I am opening this new issue as I narrowed down the issue is around timing of the 2 streams and saw the context is likely to be different from #109. I am fine to be merged with the other issue if that is more convenient.

**Comment by lsh-0** [comment]
This is a good bug report, some feedback from the maintainer(s) would be nice.

**Comment by imTHAI** [comment]
I have the same problem.
Is the project still maintained ?

**Comment by alexhexabeam** [comment]
This looks like a case for https://docs.python.org/3/library/select.html#select.select

**Comment by pkittenis** [comment]
Thanks for the report, will take a look.

**Comment by pkittenis** [comment]
Thanks for the interest and report.

The library can indeed raise Timeout too early in the case where there is no output for a given stream. This is a good test case for what is reported at #207.

However, for the above use case, the client side code should be resetting output generators if wanting to continue reading from them after timeout. This will be made automatic in future versions.

Eg:

```
from pssh.clients import ParallelSSHClient
from pssh.exceptions import Timeout

client = ParallelSSHClient(['localhost'])
read_timeout = 10
output = client.run_command(
    '/tmp/stderr_test.sh', use_pty=False, timeout=read_timeout,
    sudo=False, return_list=True)
 
for host_out in output:
    while not host_out.client.finished(host_out.channel):
        try:
           for line in host_out.stdout:
                print("Stdout: %s" % (line,))
        except Timeout:
            print("Timed out from stdout")
            client.reset_output_generators(host_out, timeout=read_timeout)
        try:
            for line in host_out.stderr:
                print("Stderr: %s" % (line,))
        except Timeout:
            print("Timed out from stdout")
            client.reset_output_generators(host_out, timeout=read_timeout)
 
client.join(output)
```

Output:
```
Timed out from stdout
Stderr: start
Timed out from stdout
Stdout: 1
Stdout: 5
Stdout: 10
Timed out from stdout
Timed out from stdout
Stdout: 15
Timed out from stdout
Stdout: 20
```

Without resetting generators output after timeouts cannot be retrieved.

The first two timeouts prior to 10sec are incorrect and due from the socket being blocked on there being no output on that stream.

This incorrectly is evaluated as the same as the socket having timed out after <timeout> seconds and raises exception.

**Comment by pkittenis** [comment]
Resolved in `1.12.1`.
